addon/index.ts

```diff
@@ -192,7 +195,38 @@ const createDecorator = (
  * @param {object?} [options={}]
  * @return {TaskProperty}
  */
-export const task = createDecorator(createTaskFromDescriptor);
+const taskDecorator = createDecorator(createTaskFromDescriptor);
+
+export function task<Args extends any[], R>(
```

Comment:
Could use docs, of course!

I *think* I'm following this; why does it require both `@task` and `= task`, though? It *seems* like it would incur extra overhead?

Reply:
_Technically_ `= task(...)` can do `@task`'s job as well, using a babel transform. This is what I explored in `ember-concurrency-typescript`. Unfortunately, I then hit https://github.com/buschtoens/ember-concurrency-typescript/issues/6 with it, which made the whole thing kinda moot.

Reply:
Got it, that's what I thought; how do we make it *not* incur runtime costs? Or is the judgment simply that they're small enough—one extra generator function allocation, it looks like?—that it doesn't really matter?

Reply:
It's not allocating an extra generator function, luckily! :)

`= task(function*() {})` / `= task({ *perform() {} })` trigger this code branch:

https://github.com/machty/ember-concurrency-decorators/blob/66830d7b5f8b4e23d51c61c5d7658620d700b72f/addon/index.ts#L219-L226

This means that in this case, `= task(...)` just passes through the generator function. So effectively, the following is equivalent:

```ts
@task
foo = task(function*() {});

// is equivalent to

@task
foo = function*() {};
```

This doesn't mean though, that we can't optimize this further with a Babel transform. ⚡️ 

```ts
@task
foo = task(function*() {});

// may be turned into

@task
*foo() {}

// or even the following to avoid _any_ extra e-c-d runtime

@task(function*() {}) foo;
``

Reply:
I need to correct myself. The following stille _does_ allocate a new generator function for every class instance:

```ts
@task
foo = function*() {}; // takes the `desc.initializer` path
```

While this one does _not_ allocate a new generator function for every class instance:

```ts
@task
foo*() {} // takes the `desc.value` path
```

https://github.com/machty/ember-concurrency-decorators/blob/5abc1d5227aeb429269109b84582644c6e3f08df/addon/index.ts#L63-L73